sql/inspect: introduce new INSPECT processor into DistSQL flow

spilchen opened the pull request:
This change integrates a new INSPECT processor into the DistSQL framework. The job coordinator for INSPECT now builds a distributed plan across the key spans of a single table (currently only supports one table). The plan is executed across nodes based on the partitioning of the table, invoking the new INSPECT processor on each participating node.

At present, the processor is a no-op that simply returns success, laying groundwork for future changes.

Informs: #148683
Epic: CRDB-30356
Release note: none

yuzefovich reviewed (commented):
Nice work! I have some minor suggestions.

> Reviewed 17 of 17 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-:-OUcIsJb5p3KAWcU7bYp:b-dq5yfx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @rafiss)

___
*[`pkg/sql/logictest/testdata/logic_test/inspect` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUcIgV_7WsRB88OTqcN:-OUcIgV_7WsRB88OTqcO:b-93e004) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/logictest/testdata/logic_test/inspect#L1)):*
> ```
> # LogicTest: !fakedist !fakedist-vec-off !fakedist-disk
> ```

nit: why are we skipping these configs? Do we plan to eventually support them? Perhaps leave a TODO then.
___
*[`pkg/sql/inspect/inspect_processor.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUcHWpH3hlLUToRSTmY:-OUcHWpH3hlLUToRSTmZ:b-50zeu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/inspect/inspect_processor.go#L25)):*
> ```Go
> 
> var _ execinfra.Processor = (*inspectProcessor)(nil)
> var _ execinfra.RowSource = (*inspectProcessor)(nil)
> ```

nit: I don't think we need to implement `RowSource` interface. In fact, the current implementation of that interface seems broken - there is no work done in `Start` and `Next` methods, so I think we should simply implement `Processor` interface and remove redundant methods.

(For context, most processors (at least owned by Queries) implement `RowSource` interface and do not explicitly implement `Processor` interface. This is achieved by embedding `ProcessorBase` helper which has the default implementation of `Processor` interface methods. Whenever a processor defines `Run` method itself, it doesn't need to become `RowSource` and does not need to embed `ProcessorBase`.)
___
*[`pkg/sql/execinfrapb/BUILD.bazel` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUc3M4N9h6cqFDKt6BC:-OUc3M4O9oLiv-priiAz:bvupsyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/execinfrapb/BUILD.bazel#L85)):*
> ```Starlark
>         "processors_bulk_io.proto",
>         "processors_changefeeds.proto",
>         "processors_inspect.proto",
> ```

nit: let's also mark this proto as owned by Foundations in CODEOWNERS.
___
*[`pkg/sql/execinfrapb/processors_inspect.proto` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUc3liSDfysXSlOYD_v:-OUc3liSDfysXSlOYD_w:b6a6nyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/execinfrapb/processors_inspect.proto#L18)):*
> ```proto
> 
> import "gogoproto/gogo.proto";
> import "google/protobuf/duration.proto";
> ```

nit: looks like this import is not used.
___
*[`pkg/sql/execinfrapb/processors_inspect.proto` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUc3pl61IG5goWCyoEP:-OUc3pl61IG5goWCyoEQ:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/execinfrapb/processors_inspect.proto#L23)):*
> ```proto
> 
> message InspectSpec {
>   // JobID of the job that ran the inspect processor
> ```

nit: missing period.
___
*[`pkg/sql/inspect/inspect_job.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUcGwgh1DR3zpGHIkXZ:-OUcGwgh1DR3zpGHIkX_:bbl6jww) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/inspect/inspect_job.go#L135)):*
> ```Go
> 	}
> 
> 	sqlInstanceIDToSpec := make(map[base.SQLInstanceID]*execinfrapb.InspectSpec, len(spanPartitions))
> ```

nit: this `map` seems redundant since we only use it once to iterate over, so why not iterate over `spanPartitions` directly when populating the core placements?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149583)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler, @rafiss, and @yuzefovich)

___
*[`pkg/sql/inspect/inspect_processor.go` line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUcHWpH3hlLUToRSTmY:-OUeyhKZ8obI104dEYqD:b-rp61jw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/inspect/inspect_processor.go#L25)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I don't think we need to implement `RowSource` interface. In fact, the current implementation of that interface seems broken - there is no work done in `Start` and `Next` methods, so I think we should simply implement `Processor` interface and remove redundant methods.

(For context, most processors (at least owned by Queries) implement `RowSource` interface and do not explicitly implement `Processor` interface. This is achieved by embedding `ProcessorBase` helper which has the default implementation of `Processor` interface methods. Whenever a processor defines `Run` method itself, it doesn't need to become `RowSource` and does not need to embed `ProcessorBase`.)
</blockquote></details>

Thanks for the tip. Yeah, I ended up copying too much from an existing processor. So, this one helped a lot.
___
*[`pkg/sql/logictest/testdata/logic_test/inspect` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUcIgV_7WsRB88OTqcN:-OUf4hFW4HHyWttOvhrR:by7w2wq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/logictest/testdata/logic_test/inspect#L1)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why are we skipping these configs? Do we plan to eventually support them? Perhaps leave a TODO then.
</blockquote></details>

I found that when they were enabled, the INSPECT job didn't play nice. It hit a SIGSEGV.
```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2d72b0f]

goroutine 7625 [running]:
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).DB(...)
    pkg/kv/txn.go:272
github.com/cockroachdb/cockroach/pkg/sql/physicalplan.(*fakeSpanResolver).NewSpanResolverIterator(0xc002592bd0, 0x0, {0xa99c000?, 0xafe8840?})
    pkg/sql/physicalplan/fake_span_resolver.go:69 +0x2f
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).NewPlanningCtxWithOracle(0xc003c96c60, {0x7021818, 0xc00b43ca50}, 0xc002e07548, 0x0, 0x0, 0x1, {0x0, 0x0}, {{0x0, ...}})
    pkg/sql/distsql_physical_planner.go:5481 +0x22f
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).setupAllNodesPlanningSystem(0xc003c96c60, {0x7021818, 0xc00b43ca50}, 0x0?, 0xc00264cc08, {0x0?, 0x0?}, {{0x0, 0x0, 0x0}})
    pkg/sql/distsql_plan_bulk.go:59 +0x89
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).SetupAllNodesPlanningWithOracle(0xc003c96c60, {0x7021818, 0xc00b43ca50}, 0xc002e07548, 0xc00264cc08, {0x0, 0x0}, {{0x0, 0x0, 0x0}})
    pkg/sql/distsql_plan_bulk.go:44 +0xd8
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).SetupAllNodesPlanning(...)
    pkg/sql/distsql_plan_bulk.go:26
github.com/cockroachdb/cockroach/pkg/sql/inspect.(*inspectResumer).planInspectProcessors(0xc00338f0c8, {0x7021818, 0xc00b43ca50}, {0x707f190, 0xc00332fbb0}, {0xc0094fa990, 0x1, 0x1})
    pkg/sql/inspect/inspect_job.go:116 +0x118
github.com/cockroachdb/cockroach/pkg/sql/inspect.(*inspectResumer).Resume(0xc00338f0c8, {0x7021818, 0xc00b43ca50}, {0x5fe7940, 0xc00332fbb0})
    pkg/sql/inspect/inspect_job.go:54 +0x15b

```

It traps trying to dereference `txn`, which is nil, here:
https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/physicalplan/fake_span_resolver.go#L69

The nil `txn` comes from here:
https://github.com/cockroachdb/cockroach/blob/a9e1061666377154cb0dcb6f67b6bcdec9278168/pkg/sql/distsql_plan_bulk.go#L60

So, it would appear that `SetupAllNodesPlanning` cannot be used in the fakedist context without some changes.

Any ideas?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 4 of 4 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149583)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler, @rafiss, and @spilchen)

___
*[`pkg/sql/logictest/testdata/logic_test/inspect` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUcIgV_7WsRB88OTqcN:-OUfXHt068_vQVLHWdN7:bpicsbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/logictest/testdata/logic_test/inspect#L1)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I found that when they were enabled, the INSPECT job didn't play nice. It hit a SIGSEGV.
```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2d72b0f]

goroutine 7625 [running]:
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).DB(...)
    pkg/kv/txn.go:272
github.com/cockroachdb/cockroach/pkg/sql/physicalplan.(*fakeSpanResolver).NewSpanResolverIterator(0xc002592bd0, 0x0, {0xa99c000?, 0xafe8840?})
    pkg/sql/physicalplan/fake_span_resolver.go:69 +0x2f
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).NewPlanningCtxWithOracle(0xc003c96c60, {0x7021818, 0xc00b43ca50}, 0xc002e07548, 0x0, 0x0, 0x1, {0x0, 0x0}, {{0x0, ...}})
    pkg/sql/distsql_physical_planner.go:5481 +0x22f
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).setupAllNodesPlanningSystem(0xc003c96c60, {0x7021818, 0xc00b43ca50}, 0x0?, 0xc00264cc08, {0x0?, 0x0?}, {{0x0, 0x0, 0x0}})
    pkg/sql/distsql_plan_bulk.go:59 +0x89
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).SetupAllNodesPlanningWithOracle(0xc003c96c60, {0x7021818, 0xc00b43ca50}, 0xc002e07548, 0xc00264cc08, {0x0, 0x0}, {{0x0, 0x0, 0x0}})
    pkg/sql/distsql_plan_bulk.go:44 +0xd8
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).SetupAllNodesPlanning(...)
    pkg/sql/distsql_plan_bulk.go:26
github.com/cockroachdb/cockroach/pkg/sql/inspect.(*inspectResumer).planInspectProcessors(0xc00338f0c8, {0x7021818, 0xc00b43ca50}, {0x707f190, 0xc00332fbb0}, {0xc0094fa990, 0x1, 0x1})
    pkg/sql/inspect/inspect_job.go:116 +0x118
github.com/cockroachdb/cockroach/pkg/sql/inspect.(*inspectResumer).Resume(0xc00338f0c8, {0x7021818, 0xc00b43ca50}, {0x5fe7940, 0xc00332fbb0})
    pkg/sql/inspect/inspect_job.go:54 +0x15b

```

It traps trying to dereference `txn`, which is nil, here:
https://github.com/cockroachdb/cockroach/blob/9f61a38b670e57873509568f67bb665e50cd1935/pkg/sql/physicalplan/fake_span_resolver.go#L69

The nil `txn` comes from here:
https://github.com/cockroachdb/cockroach/blob/a9e1061666377154cb0dcb6f67b6bcdec9278168/pkg/sql/distsql_plan_bulk.go#L60

So, it would appear that `SetupAllNodesPlanning` cannot be used in the fakedist context without some changes.

Any ideas?
</blockquote></details>

Ah, yes, makes sense. We generally think that that distsql flows that are created on behalf of a job do not run in an "outer" txn, which is why we're hitting this nil pointer. I think it's fine to skip `fakedist` configs then for `inspect` file (probably worth leaving a quick comment why we need to do this).

Relatedly though, we might want to create a separate `distsql_inspect` file that would run in `5node-default-configs` to get some distributed case test coverage. We usually explicitly split ranges and moved them around in those configs to get truly distributed plans.
___
*[`pkg/sql/inspect/inspect_processor.go` line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUfWiz786M3mYJpd1B9:-OUfWiz786M3mYJpd1BA:bbbz1mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b63f14836060000c35d7e879ffe52bf6a512fbb/pkg/sql/inspect/inspect_processor.go#L49)):*
> ```Go
> 		p.flowCtx.Cfg.TestingKnobs.ProcessorNoTracingSpan
> 	if !noSpan {
> 		ctx, _ = execinfra.ProcessorSpan(ctx, p.flowCtx, "inspect", p.processorID)
> ```

Now that we no longer embed `ProcessorBase`, we need to do `defer span.Finish()` explicitly. (Previously, this was done in `ProcessorBaseNoHelper.InternalClose`.)
___
*[`pkg/sql/inspect/inspect_job.go` line 137 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUfWErT3UBeDKwOlYJR:-OUfWErU8PP5MCDwkU1y:b-w1txg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b63f14836060000c35d7e879ffe52bf6a512fbb/pkg/sql/inspect/inspect_job.go#L137)):*
> ```Go
> 	processorCorePlacements := make([]physicalplan.ProcessorCorePlacement, len(spanPartitions))
> 	i := 0
> 	for _, spanPartition := range spanPartitions {
> ```

nit: we can define `i` here instead of `_`.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: introduce new INSPECT processor into DistSQL flow

This change integrates a new INSPECT processor into the DistSQL framework. The
job coordinator for INSPECT now builds a distributed plan across the key spans
of a single table (currently only supports one table). The plan is executed
across nodes based on the partitioning of the table, invoking the new INSPECT
processor on each participating node.

At present, the processor is a no-op that simply returns success, laying
groundwork for future enhancements.

Informs: #148683
Epic: CRDB-30356
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149583)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler, @rafiss, and @yuzefovich)

___
*[`pkg/sql/inspect/inspect_processor.go` line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUfWiz786M3mYJpd1B9:-OUj7Is3ESRIKvYhBqxF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b63f14836060000c35d7e879ffe52bf6a512fbb/pkg/sql/inspect/inspect_processor.go#L49)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Now that we no longer embed `ProcessorBase`, we need to do `defer span.Finish()` explicitly. (Previously, this was done in `ProcessorBaseNoHelper.InternalClose`.)
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/inspect` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-OUcIgV_7WsRB88OTqcN:-OUjEeWC-YVsPiBUun4A:b-xsbl4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc842d82c8dabfc84546a806357d0641c35acbd/pkg/sql/logictest/testdata/logic_test/inspect#L1)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ah, yes, makes sense. We generally think that that distsql flows that are created on behalf of a job do not run in an "outer" txn, which is why we're hitting this nil pointer. I think it's fine to skip `fakedist` configs then for `inspect` file (probably worth leaving a quick comment why we need to do this).

Relatedly though, we might want to create a separate `distsql_inspect` file that would run in `5node-default-configs` to get some distributed case test coverage. We usually explicitly split ranges and moved them around in those configs to get truly distributed plans.
</blockquote></details>

Good idea. I added the new testcase.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149583#-:-OUk7ksy5LPHaMff1NLh:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @msbutler and @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!
